hazelcast/src/main/java/com/hazelcast/spi/impl/AbstractInvocationFuture.java

```diff
@@ -228,63 +674,402 @@ private void unblockAll(Object waiter, Executor executor) {
+                                         @Nonnull CompletableFuture<Void> future) {
+        final Object value = resolve(state);
+        // in case this future is completed exceptionally, the result is also exceptionally completed
+        // without checking whether otherFuture is completed or not
+        if (cascadeException(value, future)) {
+            return;
+        }
+        if (!otherFuture.isDone()) {
+            // register on other future as waiter and return
+            otherFuture.whenCompleteAsync((u, t) -> {
+                if (t != null) {
+                    future.completeExceptionally(t);
+                }
+                try {
+                    action.accept((V) value, u);
+                    future.complete(null);
+                } catch (Throwable e) {
+                    future.completeExceptionally(e);
+                }
+            }, executor);
+            return;
+        }
+        // both futures are done
+        if (otherFuture.isCompletedExceptionally()) {
+            otherFuture.whenComplete((v, t) -> {
+                future.completeExceptionally(t);
+            });
+            return;
+        }
+        U otherValue = otherFuture.join();
+        executor.execute(() -> {
+            try {
+                action.accept((V) value, otherValue);
+                future.complete(null);
+            } catch (Throwable t) {
+                future.completeExceptionally(t);
+            }
+        });
+    }
 
+    private void unblockRunAfterBoth(@Nonnull CompletableFuture<?> otherFuture,
+                                       @Nonnull final Runnable action,
+                                       @Nonnull Executor executor,
+                                       @Nonnull CompletableFuture<Void> future) {
+        Object result = resolve(state);
+        // in case this future is completed exceptionally, the result is also exceptionally completed
+        // without checking whether otherFuture is completed or not
+        if (cascadeException(result, future)) {
+            return;
+        }
+        if (!otherFuture.isDone()) {
+            // register on other future as waiter and return
+            otherFuture.whenCompleteAsync((u, t) -> {
+                if (t != null) {
+                    future.completeExceptionally(t);
+                }
+                try {
+                    action.run();
+                    future.complete(null);
+                } catch (Throwable e) {
+                    future.completeExceptionally(e);
+                }
+            }, executor);
+            return;
+        }
+        // both futures are done
+        if (otherFuture.isCompletedExceptionally()) {
+            otherFuture.whenComplete((v, t) -> {
+                future.completeExceptionally(t);
             });
-        } catch (RejectedExecutionException e) {
-            callback.onFailure(wrapToInstanceNotActiveException(e));
```

Comment:
This exception transformation was added in #15574 in `unblock(ExecutionCallback)` method which is now obsolete. It still has to be applied on execution of other user supplied computation stages.